File: backend/src/main/kotlin/app/meals/storage/FamilyInvitationRepository.kt
package app.meals.storage

import java.sql.Connection
import java.time.Instant
import java.util.UUID

data class FamilyInvitationRow(
    val id: UUID,
    val familyId: UUID,
    val inviterUserId: UUID,
    val inviteeEmail: String,
    val token: String,
    val status: String,
    val expiresAt: Instant,
)

object FamilyInvitationRepository {
    const val STATUS_PENDING = "pending"
    const val STATUS_ACCEPTED = "accepted"

    fun insert(
        conn: Connection,
        familyId: UUID,
        inviterUserId: UUID,
        inviteeEmail: String,
        token: String,
        expiresAt: Instant,
    ): UUID {
        conn.prepareStatement(
            """
            INSERT INTO family_invitations (family_id, inviter_user_id, invitee_email, token, status, expires_at)
            VALUES (?, ?, ?, ?, ?, ?)
            RETURNING id
            """.trimIndent()
        ).use { ps ->
            ps.setObject(1, familyId)
            ps.setObject(2, inviterUserId)
            ps.setString(3, inviteeEmail)
            ps.setString(4, token)
            ps.setString(5, STATUS_PENDING)
            ps.setObject(6, java.sql.Timestamp.from(expiresAt))
            return ps.executeQuery().use { rs ->
                check(rs.next())
                UUID.fromString(rs.getString("id"))
            }
        }
    }

    fun findByToken(token: String): FamilyInvitationRow? {
        return DatabaseFactory.query { conn -> findByTokenConn(conn, token) }
    }

    fun findByTokenConn(conn: Connection, token: String): FamilyInvitationRow? {
        return conn.prepareStatement(
            """
            SELECT id, family_id, inviter_user_id, invitee_email, token, status, expires_at
            FROM family_invitations WHERE token = ?
            """.trimIndent()
        ).use { ps ->
            ps.setString(1, token)
            ps.executeQuery().use { rs ->
                if (!rs.next()) null
                else mapRow(rs)
            }
        }
    }

    fun listPendingForFamily(familyId: UUID): List<FamilyInvitationRow> {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement(
                """
                SELECT id, family_id, inviter_user_id, invitee_email, token, status, expires_at
                FROM family_invitations
                WHERE family_id = ? AND status = ?
                ORDER BY created_at DESC
                """.trimIndent()
            ).use { ps ->
                ps.setObject(1, familyId)
                ps.setString(2, STATUS_PENDING)
                ps.executeQuery().use { rs ->
                    buildList {
                        while (rs.next()) add(mapRow(rs))
                    }
                }
            }
        }
    }

    fun countPendingForFamily(familyId: UUID): Int {
        return DatabaseFactory.query { conn ->
            conn.prepareStatement(
                """
                SELECT COUNT(*) FROM family_invitations
                WHERE family_id = ? AND status = ? AND expires_at > now()
                """.trimIndent()
            ).use { ps ->
                ps.setObject(1, familyId)
                ps.setString(2, STATUS_PENDING)
                ps.executeQuery().use { rs ->
                    check(rs.next())
                    rs.getInt(1)
                }
            }
        }
    }

    fun markAccepted(conn: Connection, id: UUID): Boolean {
        conn.prepareStatement(
            "UPDATE family_invitations SET status = ? WHERE id = ? AND status = ?"
        ).use { ps ->
            ps.setString(1, STATUS_ACCEPTED)
            ps.setObject(2, id)
            ps.setString(3, STATUS_PENDING)
            return ps.executeUpdate() > 0
        }
    }

    private fun mapRow(rs: java.sql.ResultSet): FamilyInvitationRow {
        return FamilyInvitationRow(
            id = UUID.fromString(rs.getString("id")),
            familyId = UUID.fromString(rs.getString("family_id")),
            inviterUserId = UUID.fromString(rs.getString("inviter_user_id")),
            inviteeEmail = rs.getString("invitee_email"),
            token = rs.getString("token"),
            status = rs.getString("status"),
            expiresAt = rs.getTimestamp("expires_at").toInstant(),
        )
    }
}

File: backend/src/storage/familyInvitationRepository.ts
import type { Db } from "../db/kysely.js";

export const INVITE_STATUS_PENDING = "pending";
export const INVITE_STATUS_ACCEPTED = "accepted";

export interface FamilyInvitationRow {
  id: string;
  family_id: string;
  inviter_user_id: string;
  invitee_email: string;
  token: string;
  status: string;
  expires_at: Date;
}

const COLUMNS = [
  "id",
  "family_id",
  "inviter_user_id",
  "invitee_email",
  "token",
  "status",
  "expires_at",
] as const;

export class FamilyInvitationRepository {
  constructor(private readonly db: Db) {}

  async insert(
    familyId: string,
    inviterUserId: string,
    inviteeEmail: string,
    token: string,
    expiresAt: Date,
  ): Promise<string> {
    const row = await this.db
      .insertInto("family_invitations")
      .values({
        family_id: familyId,
        inviter_user_id: inviterUserId,
        invitee_email: inviteeEmail,
        token,
        status: INVITE_STATUS_PENDING,
        expires_at: expiresAt,
      })
      .returning("id")
      .executeTakeFirstOrThrow();
    return row.id;
  }

  findByToken(token: string): Promise<FamilyInvitationRow | undefined> {
    return this.db
      .selectFrom("family_invitations")
      .select(COLUMNS)
      .where("token", "=", token)
      .executeTakeFirst();
  }

  /** All pending invites for the family (regardless of expiry), newest first. */
  listPendingForFamily(familyId: string): Promise<FamilyInvitationRow[]> {
    return this.db
      .selectFrom("family_invitations")
      .select(COLUMNS)
      .where("family_id", "=", familyId)
      .where("status", "=", INVITE_STATUS_PENDING)
      .orderBy("created_at", "desc")
      .execute();
  }

  /** Count pending invites that have not yet expired. */
  async countPendingForFamily(familyId: string): Promise<number> {
    const row = await this.db
      .selectFrom("family_invitations")
      .select((eb) => eb.fn.countAll<string>().as("count"))
      .where("family_id", "=", familyId)
      .where("status", "=", INVITE_STATUS_PENDING)
      .where("expires_at", ">", new Date())
      .executeTakeFirstOrThrow();
    return Number(row.count);
  }

  async markAccepted(id: string): Promise<boolean> {
    const result = await this.db
      .updateTable("family_invitations")
      .set({ status: INVITE_STATUS_ACCEPTED })
      .where("id", "=", id)
      .where("status", "=", INVITE_STATUS_PENDING)
      .executeTakeFirstOrThrow();
    return result.numUpdatedRows > 0n;
  }
}
